src/Umbraco.Web/umbraco.presentation/content.cs

```diff
@@ -1260,5 +1281,191 @@ public static void FireBeforePublishNodeToContentCache(XmlNode node, ContentCach
         }
 
         #endregion
+
+        #region Preview
+
+        private const string PreviewCacheKey = "umbraco.content.preview";
+
+        internal void ClearPreviewXmlContent()
+        {
+            if (PreviewContent.IsSinglePreview == false) return;
+
+            var runtimeCache = ApplicationContext.Current.ApplicationCache.RuntimeCache;
+            runtimeCache.ClearCacheItem(PreviewCacheKey);
+        }
+
+        internal void ClearPreviewXmlContent(int id)
+        {
+            if (PreviewContent.IsSinglePreview == false) return;
+
+            var runtimeCache = ApplicationContext.Current.ApplicationCache.RuntimeCache;
+            var xml = runtimeCache.GetCacheItem<XmlDocument>(PreviewCacheKey);
+            if (xml == null) return;
+
+            // Check if node present, before cloning
+            var x = xml.GetElementById(id.ToString());
+            if (x == null)
+                return;
+
+            // Find the document in the xml cache
+            // The document already exists in cache, so repopulate it
+            x.ParentNode.RemoveChild(x);
+        }
+
+        internal void UpdatePreviewXmlContent(Document d)

```

Comment:
For these operations do we need to worry about locking? Currently these are not thread safe. Based on the code, i think that this would only be called by the cache refreshers which are only called by the server messenger, but there's no guarantee of that.

Reply:
No thread-safety at all, will add, should be relatively easy.

Reply:
Have pushed changes to ensure that maintaining the preview Xml document is thread safe. However: little changes were required because preview works on one unique Xml document, and therefore preview rendering is basically not thread-safe, odd things might happen at time. Nothing prevents this tree from changing while a page is rendering. This, because the mechanism that is used for the non-preview Xml document consists in cloning the document for every rendered page - which happens less often for *published* content - but we cannot do it for preview, or we'd not gain anything.